records/validation.go

```diff
@@ -3,67 +3,84 @@ package records
 import (
 	"fmt"
 	"net"
+	"strconv"
 )
 
 func validateEnabledServices(c *Config) error {
 	if !c.DNSOn && !c.HTTPOn {
 		return fmt.Errorf("Either DNS or HTTP server should be on")
 	}
 	if len(c.Masters) == 0 && c.Zk == "" {
-		return fmt.Errorf("specify mesos masters or zookeeper in config.json")
+		return fmt.Errorf("Specify Mesos masters or Zookeeper in config.json")
 	}
 	return nil
 }
 
-// validateMasters checks that each master in the list is a properly formatted host:ip pair.
+// validateMasters checks that each master in the list is a properly formatted host:port or IP:port pair.
 // duplicate masters in the list are not allowed.
 // returns nil if the masters list is empty, or else all masters in the list are valid.
 func validateMasters(ms []string) error {
-	if len(ms) == 0 {
-		return nil
-	}
-	valid := make(map[string]struct{}, len(ms))
-	for i, m := range ms {
-		h, p, err := net.SplitHostPort(m)
-		if err != nil {
-			return fmt.Errorf("illegal host:port specified for master %q", ms[i])
-		}
-		// normalize ipv6 addresses
-		if ip := net.ParseIP(h); ip != nil {
-			h = ip.String()
-			m = h + "_" + p
-		}
-		//TODO(jdef) distinguish between intended hostnames and invalid ip addresses
-		if _, found := valid[m]; found {
-			return fmt.Errorf("duplicate master specified: %v", ms[i])
-		}
-		valid[m] = struct{}{}
+	if err := validateHostPorts(ms, false, "5050", true); err != nil {
+		return fmt.Errorf("Error validating masters: %v", err)
 	}
 	return nil
 }
 
-// validateResolvers checks that each resolver in the list is a properly formatted IP address.
+// validateResolvers checks that each resolver in the list is a properly formatted IP or IP:port pair.
 // duplicate resolvers in the list are not allowed.
 // returns nil if the resolver list is empty, or else all resolvers in the list are valid.
 func validateResolvers(rs []string) error {
-	if len(rs) == 0 {
+	if err := validateHostPorts(rs, true, "53", false); err != nil {
+		return fmt.Errorf("Error validating resolvers: %v", err)
+	}
+	return nil
+}
+
+func validateHostPorts(hostPorts []string, ipRequired bool, defaultPort string, portRequired bool) error {
+	if len(hostPorts) == 0 {
```

Comment:
Just curious any reason why this short-circuit was here in the first place? I mean, I can't imagine `valid := make(map[string]struct{}, len(hostPorts))` would be so expensive to make this worth it?

CC: @jdef